src/control/stringatom.cpp

```diff
@@ -0,0 +1,75 @@
+#include <QtDebug>
+#include <QMutexLocker>
+
+#include "control/stringatom.h"
+
+#include "util/stat.h"
+#include "util/timer.h"
+
+// Static member variable definition
+QHash<QString, const QString*> StringAtom::s_stringHash;
+QMutex StringAtom::s_stringHashMutex;
+
+StringAtom::StringAtom()
+        : m_pString(NULL) {
+}
+
+StringAtom::StringAtom(const char* asciiString) {
+    QString string(asciiString);
+    initalize(string);
+}
+
+StringAtom::StringAtom(const QString& string) {
+    initalize(string);
+}
+
+StringAtom::StringAtom(const QString* pString)
+        : m_pString(pString) {
+}
+
+StringAtom::~StringAtom() {
+    // do not delete m_pString here, it is owned by the s_stringHash
+}
+
+void StringAtom::initalize(const QString& string) {
+    QMutexLocker locker(&s_stringHashMutex);
+    const QString* pString = getInner(string);
```

Comment:
You can avoid the double-hash lookup on insert by using ```QHash::operator[]``` which returns a mutable reference to the position in the hash.

```c++
QString* &pString = s_stringHashMutex[string];
if (pString == NULL) {
  pString = new QString(string);
}
```